includes/razorpay-webhook.php

```diff
@@ -110,4 +127,114 @@ protected function paymentAuthorized($data)
 
         exit;
     }
+
+    /**
+     * Currently we handle only subscription failures using this webhook
+     *
+     * @param $data
+     */
+    protected function paymentFailed($data)
+    {
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['payment']['entity']['notes']['woocommerce_order_id'];
+
+        $paymentId = $data['payload']['payment']['entity']['id'];
+
+        if (isset($data['payload']['payment']['subscription_id']) === true)
+        {
+            $this->processSubscription($orderId, $paymentId, false);
+        }
+
+        exit;
+    }
+
+    /**
+     * Handling the subscription charged webhook
+     *
+     * @param $data
+     */
+    protected function subscriptionCharged($data)
+    {
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['subscription']['entity']['notes']['woocommerce_order_id'];
+
+        $this->processSubscription($orderId);
+
+        exit;
+    }
+
+    /**
+     * Helper method used to handle all subscription processing
+     *
+     * @param $orderId
+     * @param $paymentId
+     * @param $success
+     */
+    protected function processSubscription($orderId, $paymentId, $success = true)
+    {
+        //
+        // If success is false, automatically process subscription failure
+        //
+        if ($success === false)
+        {
+            return $this->processSubscriptionFailed($orderId);
+        }
+
+        // This method is used to process the subscription's recurring payment
+        $wcSubscription = wcs_get_subscriptions_for_order($orderId);
+
+        $subscriptionId = get_post_meta($orderId, 'razorpay_subscription_id')[0];
+
+        $api = $this->razorpay->getRazorpayApiInstance();
+
+        $subscription = $api->subscription->fetch($subscriptionId);
+
+        $this->processSubscriptionSuccess($wcSubscription, $subscription, $paymentId);
+
+        exit;
+    }
+
+    /**
+     * In the case of successful payment, we mark the subscription successful
+     *
+     * @param $wcSubscription
+     * @param $subscription
+     */
+    protected function processSubscriptionSuccess($wcSubscription, $subscription, $paymentId)
+    {
+        $wcSubscriptionId = array_keys($wcSubscription)[0];
+
+        // We will only process one subscription per order
+        $wcSubscription = array_values($wcSubscription)[0];
+
+        // The subscription is completely paid for
+        if ($wcSubscription->get_completed_payment_count() === $subscription->total_count)
```

Comment:
save the output of get_completed_payment_count() in a variable. [WC seems to be caching it](https://github.com/wp-premium/woocommerce-subscriptions/blob/master/includes/class-wc-subscription.php#L712), but I'd rather still not call it twice.